### ISSUE TITLE
Question about how data binding/abstract syntax tree works

### ISSUE THREAD

**Issue by kennygitchen** [ractivejs/ractive#2363]
hi,
Forgive me if I am not making any sense here.

I am using legacy version of Ractive (need to support ie8) and I have a situation that after Ractive rendered the template. I need to re-attach an existing dom element to a new parent at the same time need the bindings to work. 

More interestingly, the new parent is outside the Ractive's scope (as not a child of the element, that is passed when create a Ractive instance). 

Here is a simple example to explain what I am doing:

``` html
<div id="newParent">
</div>
<div id="existParent">
</div> 
```

``` JavaScript
var ractive = new Ractive( { 
  el: $('#existParent'),
  template: "<div id="child">{{msg}}</div>",
  data : { msg : "hello test" }
} );
$( '#child' ).appendTo( '#newParent' )
ractive.update( 'msg', 'hello moved' );
```

I have already tested and the binding seems to be working (in Chrome) still after I moved the dom. Just wondering am I doing something that Ractive does not intentionally supports, so I might get issues in the future. 

Future more, how does Ractive keeps its binding even after I moved the dom element.

cheers, Kenny

**Comment by fskreuz** [comment]
Similar to https://github.com/ractivejs/ractive/issues/1972 and https://github.com/ractivejs/docs.ractivejs.org/issues/47 . Probably some new docs coming after 0.8 becomes mainstream.

**Comment by kennygitchen** [comment]
hi fskreuz,
So is it expected that Ractive binding will work on the child element even it is re-attached to a new element after rendered?

**Comment by fskreuz** [comment]
@kennygitchen 

> I don't seem to be able to find the how Ractive works internally or related information from the references that you have linked in your comment.

Because there isn't any formal documentation about how the internals work (hence those issues linked). There's only discussions here and there but nothing really definitive at the moment.

Documentation is one of the things coming soon. Just holding off until 0.8 comes out of the production line as a lot of things have changed since 0.7. :smile:

**Comment by kennygitchen** [comment]
@fskreuz 
So, we are not sure this feature that seems to work right now is intended or is supported?

**Comment by fskreuz** [comment]
Ok, I'll start over. :smile: 

> Question about how data binding/abstract syntax tree works

> how does Ractive keeps its binding even after I moved the dom element.

No definitive documentation yet.

> I need to re-attach an existing dom element to a new parent at the same time need the bindings to work.

Detaching and reattaching *is* definitely supported. There's [`ractive.detach`](http://docs.ractivejs.org/latest/ractive-detach) and [`ractive.insert`](http://docs.ractivejs.org/latest/ractive-insert)

> ```
var ractive = new Ractive( { 
  el: $('#existParent'),
  template: "<div id="child">{{msg}}</div>",
  data : { msg : "hello test" }
} );
$( '#child' ).appendTo( '#newParent' )
ractive.update( 'msg', 'hello moved' );
```

This is just an odd way of doing it though. And yes, you're right. This method *may* work this way like you discovered, but it's probably not intended. Use [`ractive.detach`](http://docs.ractivejs.org/latest/ractive-detach) and [`ractive.insert`](http://docs.ractivejs.org/latest/ractive-insert) instead.

**Comment by kennygitchen** [comment]
@fskreuz 
Thanks for the info, this indeed very helpful.

**Comment by sabob** [comment]
This is a great question.

I'm not a ractive expert but did some debugging and will try and explain why moving the dom works  in your example and why it is probably a bad idea ;-)

Given this:
```
var ractive = new Ractive( { 
  el: $('#existParent'),
  template: "<div id="child">{{msg}}</div>",
  data : { msg : "hello test" }
} );
$( '#child' ).appendTo( '#newParent' )
ractive.update( 'msg', 'hello moved' );
```

ractive will parse the template into an AST as described here:

https://github.com/ractivejs/template-spec

Based on the AST, ractive will build a virtual dom consisting of fragments, items, expressions etc. and "render" that virtual dom to the browser DOM. "Render" means creating and inserting  browser DOM nodes based on the AST. For example:

mustaches in the template becomes [Interpolator](https://github.com/ractivejs/ractive/blob/bd55c5e4880b78dc88a9600653a242b56c281948/src/view/items/Interpolator.js#L7) instances in the virtual dom. Interpolators are rendered to the browser DOM as [TextNodes](https://developer.mozilla.org/en-US/docs/Web/API/Document/createTextNode):

A virtual dom item has a reference to the actual browser dom node. In the case of interpolators it has a "node" property which references the TextNode in the browser DOM.


To summarize so far:

Template | Virtual Dom | Browser DOM
------------ | ----------------|------------------
{{msg}}     | Interpolator | TextNode

Textnodes have a data property to set/get it's text value.

When the keypath changes ractive will update the text node data property with the new value, for example:

```
ractive.set("msg", "hello");
```

will lead to this, internally:
```
theMsgTextNode.data = "hello";
```

Next, let's look at what happens when moving the dom without ractive knowing about it.

I'm going to change your example slightly to make it more clear what is happening. Instead of moving the "child" div, I'm going to move the content of the "child" div to a new parent, like so:

```
var node = document.getElementById("child");
var msgTextNode = node.removeChild( node.firstChild );
$( "#newParent" ).append(msgTextNode);
```

Above we move the textnode, which represents the "msg" mustache, to the "newParent" div. When we update the keypath "msg", ractive will tell the interpolator, which still has a reference to the msg textNode, to set the textNode.data to the new value.

The reason things still work is because we moved the browser DOM to a new location, but ractive still has a reference to the ntextNode we moved, so updating the keypath updates the browser DOM.

The problem though is that at this stage, according to ractive's virtual dom, the msg text node is inside the "child" div, which isn't true since we moved the content of the "child" div (the textNode) to the "newParent" div.

To verify the inconsistency between virtual dom and browser dom we can do:
```
var html = ractive.toHTML()
console.log(html);
```

this will print:
```
 "<div id='child'>hello</div>
```

where in reality it is:
```
 "<div id='newParent'>hello</div>
```

In more complex templates, manipulating the browser dom that Ractive has a virtual dom for, would probably cause problems sooner or later.

Having said all that, I often use jQuery plugins which enhances the browser dom eg. add or change browser nodes, even inside the ractive managed dom, without issues. For example, given the following ractive template:

```
<div id="parent">
  <div id="child">
  </div>
</div>
```

I'll do this:
```
$("#child").somePlugin();
```

where somePlugin() might add new nodes inside the "child" div, so the browser DOM might become:

```
<div id="parent">
  <div id="child">
    <div class="foo">
       <form>
           <input type="text" name="foo"/>
         </form>
     </div>
  </div>
</div>
```

Ractive only knows about ```<div id="child">```, not it's content. Meaning ractive won't be influenced by any changes made to the content inside "child" either. Where things could get messy is if the jQuery plugin manipulated the browser DOM, "outside" of the "child" div. For examle, removed the "child" div and attached it to another parent node. Generally though, jQuery plugins stay within the bounds of the elements they enhance.

\*Phew\*, that was a mouthful. Not sure I've everything right, so if anyone can shed more light that would be great.

Kind regards

Bob

**Comment by kennygitchen** [comment]
@sabob 
Thanks for the info, you indeed gave me some insight about Ractive and Mustaches.

In a nutshell because reference to the DOM element does not care about where the DOM element is in the document therefore, after the binding (Ractive -> Mustache -> DOM) is established, the reference (Mustaches) can always find the DOM element that it needs to update. 

@fskreuz 
Unless, I am confused, the ractive.insert() and ractifve.detach() method are used for insert/detach ractive object to DOM. Not actually detach/insert particular DOM element that I would like to move while keeps the binding and everything else to work as per normal. The simple example that I have posted does not demonstrate exactly what I needed. 

I totally understood my situation is not a usual case and thanks for all the time you have spent with me.

cheers, Kenny

**Comment by fskreuz** [comment]
@kennygitchen The way you mentioned it is like you want to *move a specific element under a component to another location*, and not the entire component DOM. If so, then I guess that's out of bounds to any method I know of. And yes, `ractive.insert` and `ractive.detach` work on the entire component.

Removing elements from their components may have unintended side effects, especially in rendering. You might want to explain the full use case here so we might find other ways. It might even be simpler than expected. :smile: 

Reopening to invite more people.
